timemachine/fe/single_topology.py

```diff
@@ -65,14 +65,19 @@ def setup_dummy_interactions_from_ff(ff, mol, dummy_group, root_anchor_atom, nbr
     bond_params, hb = top.parameterize_harmonic_bond(ff.hb_handle.params)
     angle_params, ha = top.parameterize_harmonic_angle(ff.ha_handle.params)
     improper_params, it = top.parameterize_improper_torsion(ff.it_handle.params)
+    chiral_atom_potential, _ = top.setup_chiral_restraints(chiral_atom_k)
+    chiral_atom_idxs = chiral_atom_potential.potential.idxs
+    chiral_atom_params = chiral_atom_potential.params
 
     return setup_dummy_interactions(
         hb.idxs,
         bond_params,
         ha.idxs,
         angle_params,
-        improper_params,
         it.idxs,
+        improper_params,
```

Comment:
@badisa how the fuck did this ever work?

Reply:
This 'works' because we always drop the impropers on the floor (and it seems like we still do, or we just don't exercise this in our tests). https://github.com/proteneer/timemachine/blob/master/timemachine/fe/single_topology.py#L166 The float values will never pass this check, and even with the change it seems that the output for the improper torsions is an empty array. Do we have a test case where we expect improper torsion parameters for the dummy groups?

Reply:
Thanks for shedding light on that.

>Do we have a test case where we expect improper torsion parameters for the dummy groups?

Should be able to cook one up, will look into it. I guess this should be triggered when any of these patterns are matched within a single dummy group:
https://github.com/proteneer/timemachine/blob/0b48a62def3b89712647be9e6132d434904d9c2f/timemachine/ff/params/smirnoff_2_1_0_ccc.py#L415-L421

Reply:
Sorry jumped the gun saying we don't test it (machine stopped responded and had to restart).     `test_hif2a_end_state_stability` does exercise this code and constructs dummy idxs for improper torsions. We should probably run the nightly tests for this PR before merging.

Reply:
>Should be able to cook one up, will look into it.

A small test case

```python
import numpy as np

from rdkit import Chem
from rdkit.Chem import AllChem

from timemachine.ff import Forcefield
from timemachine.fe.single_topology import SingleTopology

ff = Forcefield.load_default()

# mol that contains a match for any smarts in default ImproperTorsionHandler
smi = 'C=NC'
mol_a = Chem.AddHs(Chem.MolFromSmiles(smi))
AllChem.EmbedMolecule(mol_a)
mol_b = Chem.AddHs(Chem.MolFromSmiles(smi))
AllChem.EmbedMolecule(mol_b)

# de-map atoms associated with improper torsion terms
_, it_idxs = ff.it_handle.parameterize(mol_a)
assert len(it_idxs) > 0
core_idxs = list(set(range(mol_a.GetNumAtoms())).difference(set(it_idxs.flatten())))
core = np.array(list(zip(core_idxs, core_idxs)))
assert len(core) > 0

# construct SingleTopology instance
st = SingleTopology(mol_a, mol_b, core, ff)
```